code/go/0chain.net/sharder/blockstore/bwr.go

```diff
@@ -0,0 +1,188 @@
+package blockstore
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"0chain.net/core/common"
+	"0chain.net/core/ememorystore"
+	"0chain.net/core/viper"
+	"github.com/0chain/gorocksdb"
+)
+
+type WhichTier uint8
+
+const (
+	DiskTier        WhichTier = 2 // Block is in disk only
+	ColdTier        WhichTier = 4 // Block is in cold storage only
+	DiskAndColdTier WhichTier = 8 // Block is in both disk and cold storages
+)
+
+const (
+	BWRCF = "bwr" // bwr column family
+	UBRCF = "ubr" // ubr column family
+	BMR   = "bmr" // block meta record
+)
+
+var bmrDB *gorocksdb.DB
+var bwrHandle *gorocksdb.ColumnFamilyHandle
+var ubrHandle *gorocksdb.ColumnFamilyHandle
+
+type blockWhereRecord struct {
+	Hash    string    `json:"-"`
+	Tiering WhichTier `json:"tr"`
+	//For disk volume it is simple unix path. For cold storage it is "storageUrl:bucketName"
+	BlockPath string `json:"vp,omitempty"`
+	ColdPath  string `json:"cp,omitempty"`
+}
+
+func (bwr *blockWhereRecord) addOrUpdate() error {
+	data, err := json.Marshal(bwr)
+	if err != nil {
+		return err
+	}
+	wo := gorocksdb.NewDefaultWriteOptions()
+	err = bmrDB.PutCF(gorocksdb.NewDefaultWriteOptions(), bwrHandle, []byte(bwr.Hash), data)
+	wo.Destroy()
+	return err
+}
+
+func getBWR(hash string) (*blockWhereRecord, error) {
+	ro := gorocksdb.NewDefaultReadOptions()
+	defer ro.Destroy()
+	dataS, err := bmrDB.GetCF(ro, bwrHandle, []byte(hash))
+	if err != nil {
+		return nil, err
+	}
+	defer dataS.Free()
+
+	data := dataS.Data()
+	bwr := &blockWhereRecord{}
+	err = json.Unmarshal(data, bwr)
+	if err != nil {
+		return nil, err
+	}
+	bwr.Hash = hash
+	return bwr, nil
+}
+
+/**********************************Unmoved block record***************************************/
+type unmovedBlockRecord struct {
+	Hash string `json:"h"`
+	// CreateAt duration passed from epoch date
+	CreatedAt common.Timestamp `json:"c"`
+}
+
+func (ubr *unmovedBlockRecord) Add() error {
+	k := strconv.FormatInt(int64(ubr.CreatedAt), 10)
+	wo := gorocksdb.NewDefaultWriteOptions()
+	err := bmrDB.PutCF(gorocksdb.NewDefaultWriteOptions(), ubrHandle, []byte(k), []byte(ubr.Hash))
+	wo.Destroy()
+	return err
+}
+
+func (ubr *unmovedBlockRecord) Delete() error {
+	k := strconv.FormatInt(int64(ubr.CreatedAt), 10)
+	wo := gorocksdb.NewDefaultWriteOptions()
+	err := bmrDB.DeleteCF(wo, ubrHandle, []byte(k))
+	wo.Destroy()
+	return err
+}
+
+// getUnmovedBlockRecords will return a channel where it will pass
+// all the unmoved blocks that is older than the block movement time interval
+func getUnmovedBlockRecords(maxPrefix []byte) <-chan *unmovedBlockRecord {
+	ch := make(chan *unmovedBlockRecord)
+
+	go func() {
+		defer close(ch)
+
+		ro := gorocksdb.NewDefaultReadOptions()
+		defer ro.Destroy()
+
+		it := bmrDB.NewIteratorCF(ro, ubrHandle)
+		defer it.Close()
+		for it.SeekToFirst(); it.Valid() && bytes.Compare(it.Key().Data(), maxPrefix) != 1; it.Next() {
```

Comment:
What does the `bytes.Compare() !=  1` expect, can we write it in more general way, or comment the logic behind it.